### ISSUE TITLE
Proposal: Breadcrumb control for navigation trail 

### ISSUE THREAD

**Issue by kat-y** [microsoft/microsoft-ui-xaml#3326]
# Proposal: Create Breadcrumb Control

A  [spec](https://github.com/microsoft/microsoft-ui-xaml-specs/blob/user/kat-y/breadcrumb/active/Breadcrumb/Breadcrumb.md) has been opened with [a PR](https://github.com/microsoft/microsoft-ui-xaml-specs/pull/102) for discussing the Breadcrumb spec. Feel free to continue general discussion in this proposal! 
## Summary

A Breadcrumb control can provide a persisting 'trail' that shows the user's navigation path. 

## Rationale

* Breadcrumb is a common UI that allows users to see the path of where they have navigated, especially when 2+ layers deep
* Breadcrumb scenarios exist in various places, including settings and file/folder structure navigation
* Creating a Fluent Breadcrumb control would increase consistency and accessibility

## Scope

| Capability | Priority |
| :---------- | :------- |
| Provide a persisting trail of the navigation path for a user when 2+ layers away from "home" | Must |
| Allow user to go back to points in their navigation path | Must |
| "Compress" the points when resized such that the full path is too long | Could |

## Important Notes

![image](https://user-images.githubusercontent.com/60154020/94196109-c0079e00-fe68-11ea-8abc-1f97023c128d.png)

### Breadcrumb controls are in:
* [Fluent UI for web](https://developer.microsoft.com/en-us/fluentui#/controls/web/breadcrumb)
* [Fabric JS/Fabric React](https://developer.microsoft.com/en-us/fabric-js/components/breadcrumb/breadcrumb)
* [UWP Breadcrumb repo](https://github.com/goedware/UWP-Breadcrumb-Control#:~:text=Join%20GitHub%20today%20%20%20%20Name%20,4%20years%20ago%20%203%20more%20rows%20) by @goedware
* [DevExpress Breadcrumb ](https://docs.devexpress.com/WPF/401002/controls-and-libraries/navigation-controls/breadcrumb-control)
* [Telerik UI RadBreadcrumb ](https://docs.telerik.com/devtools/wpf/controls/radbreadcrumb/getting-started)

## Open Questions

- Are there app scenarios other than settings and nested folders that would use a Breadcrumb control?
- The DevExpress and Telerik UI WPF Breadcrumbs control have flyout menus to select a different path to jump to, is this a feature apps using WinUI would want, and in what scenarios? 
![image](https://user-images.githubusercontent.com/60154020/94196940-e8dc6300-fe69-11ea-81e9-8f34dde61aae.png)
- The Fluent UI Breadcrumb's compression behavior cuts crumbs from the middle of the path. If this is implemented as a default behavior, what parameters should a developer have access to to change how the path is compressed? 
![image](https://user-images.githubusercontent.com/60154020/94196917-dfeb9180-fe69-11ea-91d4-02115fe83176.png)
- As it is possible to build a Breadcrumb using existing pieces from WinUI and chevron icons, would a 'style example' in XAML Controls Gallery (or somewhere else open-source) rather than a completely new control help developers to build Breadcrumbs with consistency and Fluent design? This could be built using ListView, ItemsRepeater, ComboBox or other existing controls.

**Comment by duke7553** [comment]
@kat-y We could really use a control like this in Files. 

Our current solution involving the use of a ListView is poorly-suited, and users complain of awkwardly small touch targets for the expand button. I'm glad to see a proposal for standardizing this behavior!

![image](https://user-images.githubusercontent.com/20365014/94210112-e430a280-fe9b-11ea-93de-835df0588b25.png)

**Comment by kat-y** [comment]
> Our current solution involving the use of a ListView is poorly-suited, and users complain of awkwardly small touch targets for the expand button. 

@duke7553 could you elaborate here - what is the touch target for the expand button? 
Is this scenario related to this open question? 
 - The DevExpress and Telerik UI WPF Breadcrumbs control have flyout menus to select a different path to jump to, is this a feature apps using WinUI would want, and in what scenarios?

**Comment by mdtauk** [comment]
Any File Explorer replacement should include the context/drop down menu as a UI feature - but what populates that menu should probably be left to the developer.

I would say that going up a level, as well as handling of Forward and Backward navigation, should be made as simple as possible to the developer - so maybe including a back/forward stack that can be hooked up to buttons.

**Comment by kat-y** [comment]
> I would say that going up a level, as well as handling of Forward and Backward navigation, should be made as simple as possible to the developer - so maybe including a back/forward stack that can be hooked up to buttons.

@mdtauk This sounds like separate forward/backwards buttons (often in a file navigation system) that would live next to a Breadcrumb trail - do you see this as in-scope for Breadcrumb control itself? 

Just updated the proposal to have a new open question:
- As it is possible to build a Breadcrumb using existing pieces from WinUI and chevron icons, would a 'style example' in XAML Controls Gallery (or somewhere else open-source) rather than a completely new control help developers to build Breadcrumbs with consistency and Fluent design?

**Comment by mdtauk** [comment]
> > I would say that going up a level, as well as handling of Forward and Backward navigation, should be made as simple as possible to the developer - so maybe including a back/forward stack that can be hooked up to buttons.
> 
> @mdtauk This sounds like separate forward/backwards buttons (often in a file navigation system) that would live next to a Breadcrumb trail - do you see this as in-scope for Breadcrumb control itself?
> 
> Just updated the proposal to have a new open question:
> 
> * As it is possible to build a Breadcrumb using existing pieces from WinUI and chevron icons, would a 'style example' in XAML Controls Gallery (or somewhere else open-source) rather than a completely new control help developers to build Breadcrumbs with consistency and Fluent design?

I don't see the navigation buttons being part of the control, but the control managing/maintaining navigation stacks which a developer could wire up to buttons might be something to consider.  Of course it should be for the more experienced developers to provide some response to that suggestion - they may find it too limiting relying on the control to keep track of previous navigation states/targets - and would prefer to handle that themselves.

Beyond a consistent style and behaviour pattern for the breadcrumb bar - I am not qualified to offer a definitive suggestion.

FluentUI has a discreet Breadcrumb control
https://developer.microsoft.com/en-us/fluentui#/controls/web/breadcrumb
![image](https://user-images.githubusercontent.com/7389110/94322910-cfcbd300-ff8b-11ea-9cbf-deca62bdd041.png)


And of course the most obvious example of the Windows UI using it is File Explorer (drop downs for folder choosing at the same level)
![image](https://user-images.githubusercontent.com/7389110/94322702-1d940b80-ff8b-11ea-8e61-06f60691b98e.png)
![image](https://user-images.githubusercontent.com/7389110/94322857-a1e68e80-ff8b-11ea-8ebb-47c1bdd5c9ea.png)



Google Drive has one - using a drop down on the "current/active" breadcrumb item
![image](https://user-images.githubusercontent.com/7389110/94322787-50d69a80-ff8b-11ea-913c-c75d11d165e6.png)
![image](https://user-images.githubusercontent.com/7389110/94322812-72378680-ff8b-11ea-802e-a5b886ae1a62.png)


Both the UWP versions of files apps made by Microsoft use breadcrumb bars.
![image](https://user-images.githubusercontent.com/7389110/94323076-5680b000-ff8c-11ea-8040-3757ef8c983a.png)
![image](https://thewincentral.com/wp-content/uploads/2020/03/Windows-10X-Modern-File-explorer-1-scaled.jpg)

**Comment by duke7553** [comment]
@kat-y Yes. The touch target I'm referring to would be the divider button which, when invoked, displays a flyout with other locations to navigate to. Users have told us that this dropdown is hard to invoke with touch input, so when prototyping this new control the team should keep this other input method in mind.

**Comment by kat-y** [comment]
Sounds like the flyout menu behavior is something useful to Files and other folder-navigating scenarios. 
 
>they may find it too limiting relying on the control to keep track of previous navigation states/targets - and would prefer to handle that themselves.

@mdtauk I agree with you - it's not within the scope of a Breadcrumb to track these for developers. 

Would love to hear any thoughts on the following direction for Breadcrumb:

- As it is possible to build a Breadcrumb using existing pieces from WinUI and icons, would a 'style example' in XAML Controls Gallery (or somewhere else open-source) rather than a completely new control help developers to build Breadcrumbs with consistency and Fluent design? This could be built using ListView, ItemsRepeater, ComboBox or other existing controls.

**Comment by mdtauk** [comment]
> * As it is possible to build a Breadcrumb using existing pieces from WinUI and icons, would a 'style example' in XAML Controls Gallery (or somewhere else open-source) rather than a completely new control help developers to build Breadcrumbs with consistency and Fluent design? This could be built using ListView, ItemsRepeater, ComboBox or other existing controls.
>

I think almost all controls are made up of individual elements and controls - so sure you could use a series of re-styled buttons with extra buttons - but not having it as a control means it becomes very likely individual approaches will show inconsistent behaviour and visual styling.  They may not take into account accessibility concerns.  And this would be another disparity between WinUI and FluentUI.

**Comment by zipgenius** [comment]
A breadcrumb control must definitely be included in WinUI. In my application, ZipGenius 6, I had it because Delphi had plenty of VCL components for this specific task and one of them was a perfect clone of File Explorer's one from Windows Vista/7.
I would surely use a breadcrumb bar in ZipGenius new version, which I'm developing with WinUI, .Net, C# and so on. This kind of control is absolutely needed for applications that allow users to move between locations and file managers are not the only scenarios: just imagine how could benefit a CRM software, as example.
The breadcrumb made by Goedware is good but not enough because it doesn't have flout menus attached to the dividers; however it could be a great starting point.

**Comment by kat-y** [comment]
Happy to share that this proposal has been approved for [spec writing](https://github.com/microsoft/microsoft-ui-xaml-specs/tree/user/kat-y/breadcrumb)! 

You can [see the spec with API and other details here](https://github.com/microsoft/microsoft-ui-xaml-specs/blob/user/kat-y/breadcrumb/active/Breadcrumb/Breadcrumb.md)! If you have comments or feedback on Breadcrumb, [please continue discussion in the associated PR. ](https://github.com/microsoft/microsoft-ui-xaml-specs/pull/102)

As part of discussion with the WinUI team, the work for Breadcrumb has been scoped to V1 and V2 stages of the API. In V1 Breadcrumb no chevrons are interactable to show children. V2 Breadcrumb will add functionality to have flyouts from chevrons for every node, to view children of that node. This spec focuses on V1 Breadcrumb, but keeping in mind that V1 should be positioned in a way to have V2 features smoothly added later.

**Comment by yaichenbaum** [comment]
I think it's important to take into consideration the size of the chevrons. We are currently using a custom implementation in [Files](https://github.com/files-community/Files) and users have given us feedback that the chevrons are hard to select with both touch and mouse input.
![image](https://user-images.githubusercontent.com/39923744/101540714-6cd4c000-396e-11eb-9db1-24a7ac2c7b9c.png)

**Comment by michael-hawker** [comment]
@kat-y can't this be closed now that the feature has shipped?

**Comment by yaichenbaum** [comment]
> @kat-y can't this be closed now that the feature has shipped?

Can you link the PR?

**Comment by michael-hawker** [comment]
[Original PR is here](https://github.com/microsoft/microsoft-ui-xaml/pull/3891) - looks like triage role doesn't even let you link those together either... 😥

**Comment by yaichenbaum** [comment]
The current implementation doesn't support dropdowns yet, which is the main reason we haven't switched to it for Files. Anyways, this can be tracked in a separate issue.
![image](https://user-images.githubusercontent.com/39923744/172898918-cc77990c-b88b-422a-bac9-2f13df72bcb1.png)

**Comment by michael-hawker** [comment]
@yaichenbaum yeah, I was thinking of maybe needing that too for what I wanted to do (though involves some more parsing to get the items I need, so I haven't tried yet). Was wondering if I could add a pop-up to my item's data template to handle that, though I suppose it's usually the `>` inbetween that gets clicked on to display.

Is there a separate issue tracking that specific item already?
